controllers/dexclient_controller.go

```diff
@@ -47,13 +52,143 @@ type DexClientReconciler struct {
+			// log.Info("created DexApiClient", "ca", ctls.caPEM.String(), "clientCrt", ctls.clientPEM.String(), "clientKey", ctls.clientPrivKeyPEM.String())
+			r.DexApiClient = dexApiClient
+		} else {
+			log.Info("found DexApiClient is defined")
+			info, err := r.DexApiClient.GetServerInfo(ctx)
+			if err != nil {
+				log.Info("Failed to get grpc server status:", "error", err)
+				return ctrl.Result{Requeue: true}, err
+			}
+			log.Info("DexApiClient is defined", "grpc server info:", info)
+		}
+		// Drop out
+		// return ctrl.Result{}, nil
+	default:
+		log.Info("dexClient created")
+	}
+
+	// if status is not set, set it to CREATING
+	if dexv1Client.Status.State == "" || dexv1Client.Status.State == authv1alpha1.PhaseCreating {
+		dexv1Client.Status.State = authv1alpha1.PhaseCreating
+	}
+	// Now let's make the main case distinction: implementing
+	// the state diagram CREATING -> ACTIVE or CREATING -> FAILED
+	switch dexv1Client.Status.State {
+	case authv1alpha1.PhaseCreating:
+		log.Info("Creating dex client", "name", dexv1Client.Name,
+			"redirectURIs", dexv1Client.Spec.RedirectURIs,
+			"TrustedPeers", dexv1Client.Spec.TrustedPeers,
+			"Public", dexv1Client.Spec.Public,
+			"ClientID", dexv1Client.Spec.ClientID,
+			"LogoURL", dexv1Client.Spec.LogoURL,
+			"clientSecret", dexv1Client.Spec.ClientSecret)
+
+		// Implement dex auth client creation here
+		res, err := r.DexApiClient.CreateClient(
+			ctx,
+			dexv1Client.Spec.RedirectURIs,
+			dexv1Client.Spec.TrustedPeers,
+			dexv1Client.Spec.Public,
+			dexv1Client.Spec.ClientID, // name
+			dexv1Client.Name,          // clientid
+			dexv1Client.Spec.LogoURL,
+			dexv1Client.Spec.ClientSecret, // secret
+		)
+		if err != nil {
+			log.Error(err, "Client create failed", "client", dexv1Client.Name)
+			dexv1Client.Status.State = authv1alpha1.PhaseFailed
+			dexv1Client.Status.Message = err.Error()
+		} else {
+			dexv1Client.Status.State = authv1alpha1.PhaseActive
+			log.Info("Client created", "client ID", res.GetId())
+		}
+
+	case authv1alpha1.PhaseActive:
+		// If the client is active but in the reconcile loop it's being updated.
+		log.Info("Client update", "client ID", dexv1Client.Name)
+		err := r.DexApiClient.UpdateClient(
+			ctx,
+			dexv1Client.Name, // clientid
+			dexv1Client.Spec.RedirectURIs,
+			dexv1Client.Spec.TrustedPeers,
+			dexv1Client.Spec.Public,
+			dexv1Client.Spec.ClientID,
+			dexv1Client.Spec.LogoURL,
+		)
+		if err != nil {
+			log.Error(err, "Client update failed", "client", dexv1Client.Name)
+			dexv1Client.Status.State = authv1alpha1.PhaseActiveDegraded
+			dexv1Client.Status.Message = err.Error()
+		} else {
+			log.Info("Client updated", "client ID", dexv1Client.Name)
+		}
+
+	case authv1alpha1.PhaseFailed:
+		log.Info("Client failed")
+	default:
+		// Should never reach here
+		log.Info("Got an invalid state", "state", dexv1Client.Status.State)
+		return ctrl.Result{}, nil
+	}
+	// Update the object and return
+	err := r.Update(ctx, dexv1Client)
```

Comment:
I'm not sure I'm understanding here. Shouldn't we just update the status?

Reply:
ok, will look into UpdateStatus() vs. Update() used here.

Reply:
maybe r.Status().Update ?